crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_structure.rs

```diff
@@ -125,4 +165,109 @@ mod tests {
             }
         }
     }
+    #[test]
+    fn we_can_correctly_identify_index_existence() {
+        assert!(index_exists(0));
+        assert!(!index_exists(no_index()));
+        assert!(index_exists(no_index() - 1));
+    }
+    #[test]
+    fn we_can_find_the_full_width_of_row() {
+        let width_of_row = full_width_of_row(0);
+        assert_eq!(width_of_row, 1);
+
+        let width_of_row = full_width_of_row(1);
+        assert_eq!(width_of_row, 2);
+
+        let width_of_row = full_width_of_row(4);
+        assert_eq!(width_of_row, 4);
+
+        let width_of_row = full_width_of_row(7);
+        assert_eq!(width_of_row, 8);
+
+        let width_of_row = full_width_of_row(11);
+        assert_eq!(width_of_row, 8);
+
+        let width_of_row = full_width_of_row(12);
+        assert_eq!(width_of_row, 16);
+
+        let width_of_row = full_width_of_row(13);
+        assert_eq!(width_of_row, 16);
```

Comment:
Can you do this exhaustively? I want to see that this works for rows 0..15 at least. imo, row 2 is still almost certainly an edge case.